Updating Vector*.IsHardwareAccelerated to be recursive

tannergooding opened the pull request:
This resolves https://github.com/dotnet/runtime/issues/38162

This updates the various `IsHardwareAccelerated` APIs to be recursive, just as they are for the `IsSupported` APIs. This ensures that they return the same result whether invoked directly or indirectly (such as via reflection). This ensures that the debugger and other scenarios report a consistent worldview.

tannergooding committed:
Updating Vector*.IsHardwareAccelerated to be recursive so it works in the debugger

tannergooding committed:
Adding tests validating indirect and direct invocation of get_IsHardwareAccelerated return the same value

stephentoub commented on src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector128Tests.cs:
```diff
@@ -8,6 +8,13 @@ namespace System.Runtime.Intrinsics.Tests.Vectors
+        [Fact]
+        public unsafe void Vector128IsHardwareAcceleratedTest()
+        {
+            var methodInfo = typeof(Vector128).GetMethod("get_IsHardwareAccelerated");
```
nit
```suggestion
            MethodInfo methodInfo = typeof(Vector128).GetMethod("get_IsHardwareAccelerated");
```

stephentoub commented on src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector256Tests.cs:
```diff
@@ -8,6 +8,13 @@ namespace System.Runtime.Intrinsics.Tests.Vectors
+        [Fact]
+        public unsafe void Vector256IsHardwareAcceleratedTest()
+        {
+            var methodInfo = typeof(Vector256).GetMethod("get_IsHardwareAccelerated");
```
nit
```suggestion
            MethodInfo methodInfo = typeof(Vector256).GetMethod("get_IsHardwareAccelerated");
```

stephentoub commented on src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector64Tests.cs:
```diff
@@ -8,6 +8,13 @@ namespace System.Runtime.Intrinsics.Tests.Vectors
+        [Fact]
+        public unsafe void Vector64IsHardwareAcceleratedTest()
+        {
+            var methodInfo = typeof(Vector64).GetMethod("get_IsHardwareAccelerated");
```
nit
```suggestion
            MethodInfo methodInfo = typeof(Vector64).GetMethod("get_IsHardwareAccelerated");
```

tannergooding committed:
Ensure that Vector<T>.IsHardwareAccelerated supports being recursive

tannergooding commented:
@davidwrighton, looks like `ReadyToRun` for `Vector256.IsHardwareAccelerated` is returning `false` for the indirect invocation (meaning its baked into `corelib` as `false`).

Is that expected? We don't appear to do the same for `Vector<T>` or `Avx/Avx2`

tannergooding committed:
Apply suggestions from code review

Co-authored-by: Stephen Toub <stoub@microsoft.com>

tannergooding commented on src/libraries/System.Numerics.Vectors/tests/GenericVectorTests.cs:
```diff
@@ -25,6 +25,13 @@ static GenericVectorTests()
+        [Fact]
+        public unsafe void IsHardwareAcceleratedTest()
+        {
+            var methodInfo = typeof(Vector).GetMethod("get_IsHardwareAccelerated");
```
```suggestion
            MethodInfo methodInfo = typeof(Vector).GetMethod("get_IsHardwareAccelerated");
```

tannergooding committed:
Update src/libraries/System.Numerics.Vectors/tests/GenericVectorTests.cs

tannergooding commented:
CC. @dotnet/crossgen-contrib for help with the unexpected failure above.

jeffhandley commented:
@davidwrighton Do you think there's something lurking in these unexpected failures that we need to make sure to address in 7.0?

danmoseley commented:
I think I reported this originally - we're trying to make writing vectorized code more accessible in .NET 7, and this is likely to confuse initially as it did me, so it would be nice to get it in .NET 7.

stephentoub commented:
@davidwrighton, are you still able to look at this?

danmoseley commented:
@davidwrighton ?

davidwrighton commented:
Sorry, this fell off my radar. I'm looking at it today.

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into ishardwareaccelerated-reflection

davidwrighton committed:
Detect the one legal violation of the behaves the same rule for intrinsics in CoreLib
- The functions which detect if an intrinsic is actually available are allowed to have behavior which differs based on which intrinsics are available at runtime

tannergooding committed:
Merge remote-tracking branch 'dotnet/main' into ishardwareaccelerated-reflection

tannergooding commented:
@fanyang-mono, could you assist with the changes needed for Mono here?

I went and looked at `intrinsics.c` and `simd-intrinsics.c`, but it looks like the `get_IsHardwareAccelerated` methods already have handling and similar handling to `get_IsSupported`, so its not clear what's missing and causing it to fail on the recursion rather than always creating a constant true/false node

fanyang-mono commented:
@tannergooding I will see if I have time to get to it today or tmr. Otherwise, I will forward this to another person on mono team to handle. Because I will be on vacation starting from this Friday for two weeks.

tannergooding commented:
Thanks much!

danmoseley commented:
@fanyang-mono did you manage to identify someone who could help with the mono side?

fanyang-mono commented:
@vargaz Could you help out with this while I am on vacation?

tannergooding commented:
ping @fanyang-mono, @vargaz on this.

I imagine its too late for .NET 7 at this point, but it would still be nice to get fixed

danmoseley commented:
@tannergooding is the CoreCLR change mergeable on its own?

danmoseley commented:
Also, I wouldn't rule out .NET 7 if the risk is low, as it's a goal of the release to make it easy to use vectorization in your own code. This trips you up right away.

tannergooding commented:
> @tannergooding is the CoreCLR change mergeable on its own?

Technically, yes. But there's no real point, IMO, if we aren't going to ship Corelib with the support since users will never trigger the code.

> Also, I wouldn't rule out .NET 7 if the risk is low, as it's a goal of the release to make it easy to use vectorization in your own code. This trips you up right away.

I'm not sure I'd say this is a right away scenario. Calling this via reflection or indirectly such as via a delegate is expected to be an extremely rare/niche scenario and not the normal usage pattern.

I expect the debugger case is more likely, but we really need the Mono support as well.

danmoseley commented:
Ah. Agreed, I think the debugger case is the only one that is significant.

danmoseley commented:
cc @SamMonoRT since it's potentially 7.0

vargaz committed:
[mono] Intrinsify IsHardwareAccelerated in the interpreter.

vargaz commented:
This will fix the mono interpreter failures:
https://github.com/vargaz/runtime/commit/8be6b8dbb2968d416924c49759aafcf89241f81b
Were there any other failures ?

tannergooding committed:
Merge remote-tracking branch 'dotnet/main' into ishardwareaccelerated-reflection

danmoseley commented:
relevant failures eg
```
Assert failure(PID 7492 [0x00001d44], Thread: 5828 [0x16c4]): Assertion failed '!"Unhandled must expand intrinsic, throwing PlatformNotSupportedException"' in 'System.Runtime.Intrinsics.Vector256:get_IsHardwareAccelerated():bool' during 'Importation' (IL size 6; hash 0x068ee1df; FullOpts)

File: D:\a\_work\1\s\src\coreclr\jit\importer.cpp Line: 4938
Image: C:\h\w\A5E108DA\p\corerun.exe
```
and mono failures eg JIT.HardwareIntrinsics.X86.Bmi1

tannergooding committed:
Add a fallback path for unhandled recursive intrinsics in System.Numerics

tannergooding commented:
Its actually not clear to me why CI is asserting. It doesn't assert locally at all.

However, I've added an extra safety path to see if that resolves the issue.

tannergooding committed:
Move the Vector.IsHardwareAccelerated handling into SimdAsHWIntrinsicInfo.lookupId

tannergooding commented:
Actually, realized the potential issue right as I typed that up. The path that could convert `IsHardwareAccelerated` to `true` was outside the `FEATURE_HW_INTRINSICS` path when it should've been "in" with the fallback handler always returning `false`.

tannergooding committed:
Fixinng a compiler error due to an unresolved member

tannergooding commented:
> Were there any other failures ?

@vargaz, looks like still some Mono Linux failures where its unexpectedly getting `System.PlatformNotSupportedException`.

-----

There is also a failure for RyuJIT `x64/x86 Checked no_tiered_compilation`. It's saying:
```
Assert failure(PID 7696 [0x00001e10], Thread: 5688 [0x1638]): Assertion failed '!"Unhandled must expand intrinsic, throwing PlatformNotSupportedException"' in 'System.Runtime.Intrinsics.Vector128:get_IsHardwareAccelerated():bool' during 'Importation' (IL size 6; hash 0x330ddaf5; FullOpts)

File: D:\a\_work\1\s\src\coreclr\jit\importer.cpp Line: 4938
```

@davidwrighton, this appears to be because these tests disable SSE2 and so the intrinsic returned is `NI_IsSupported_Dynamic`. I don't see any handling for this in the JIT, so what's supposed to happen for it in this recursive case?

vargaz commented:
The mono failures are also present in HEAD, so they are unrelated:
https://github.com/dotnet/runtime/runs/8369467985

tannergooding committed:
Adjust the NI_IsSupported_Dynamic checks to be NAOT only

tannergooding commented:
Failures are https://github.com/dotnet/runtime/issues/75753 and https://github.com/dotnet/runtime/issues/75606

tannergooding commented:
@danmoseley, are you still wanting this backported to .NET 7?

For `Vector64/128/256<T>` this is "new API functionality". For `Vector<T>` this has been the behavior since it was released in 2014.

There have been only a couple user reports on this, so I'm not sure if it meets the bar or not.

danmoseley commented:
Yeah now we branched, I don't think it does. Thanks for driving it though.
